Feature/dataset observation endpoint

nshumoogum opened the pull request:
### What

Update dataset API  tests to include test for `.../observations` endpoint. Also updated request headers to simulate the requests in live (more realistic) and updated version/instance documents on setup to include the headers field and to have valid values.

### How to review

Make sure extension of acceptance tests make sense and the tests pass by doing the following:

* Run `dp-auth-api-stub`
* Make sure mongo db instance is running
* Make sure neo4j is running
* Run `dp-dataset-api` on branch `feature/observations-endpoint`
* Make sure `dp-api-tests` datasetAPI acceptance tests all pass, run `cd datasetAPI; go tests ./...` in root of repository

### Who can review

Team B

nshumoogum committed:
Update dataset API tests - request headers changes

Request headers are more realistic using X-Florence Token

Also version/instance resources in setup have valid headers

nshumoogum committed:
Add dataset API observations endpoint tests

nshumoogum committed:
Uncomment out observation tests

gedge commented on datasetAPI/get_instance_test.go:
```diff
@@ -161,14 +161,23 @@ func checkResponse(datasetID, edition, instanceID, version string, response *htt
 	response.Value("dimensions").Array().Element(0).Object().Value("name").Equal("age")
-	response.Value("downloads").Object().Value("csv").Object().Value("url").String().Match("(.+)/aws/census-2017-" + version + "-csv$")
-	response.Value("downloads").Object().Value("csv").Object().Value("url").String().Match("(.+)/aws/census-2017-" + version + "-csv$")
+	response.Value("downloads").Object().Value("csv").Object().Value("href").String().Match("(.+)/aws/census-2017-" + version + "-csv$")
```
I think you can remove `(.+)` from each of these (you don't need to match `.+` or even capture `(...)` it)

gedge commented on datasetAPI/get_instance_test.go:
```diff
@@ -161,14 +161,23 @@ func checkResponse(datasetID, edition, instanceID, version string, response *htt
-	response.Value("downloads").Object().Value("csv").Object().Value("url").String().Match("(.+)/aws/census-2017-" + version + "-csv$")
+	response.Value("downloads").Object().Value("csv").Object().Value("href").String().Match("(.+)/aws/census-2017-" + version + "-csv$")
+	response.Value("downloads").Object().Value("csv").Object().Value("private").String().Match("(.+)/csv-exported/myfile.csv$")
+	response.Value("downloads").Object().Value("csv").Object().Value("public").String().Match("(.+)/csv-exported/myfile.csv$")
```
above two lines:  can we have a different `public`/`private` values to check (there may be a bug causing both to be the same, when they should differ)

sorry, lots of examples of these elsewhere...

gedge commented on datasetAPI/get_observation_test.go:
```diff
@@ -0,0 +1,306 @@
+				response.Value("dimensions").Object().Value("time").Object().Value("options").Array().Element(0).Object().Value("id").Equal("Aug-16")
+				response.Value("observation").Equal("154.6")
+				response.Value("links").Object().Value("dataset_metadata").Object().Value("href").String().Match("(.+)/datasets/" + datasetID + "/editions/" + edition + "/versions/2/metadata$")
+				response.Value("links").Object().Value("self").Object().Value("href").String().Match(".+/datasets/" + datasetID + "/editions/" + edition + "/versions/2/observations\\?aggregate=cpi1dim1S40403&geography=K02000001&time=Aug-16$") // ?aggregate=cpi1dim1S40403&geography=K02000001&time=Aug-16")
```
remove trailing comment

gedge commented on datasetAPI/get_observation_test.go:
```diff
@@ -0,0 +1,306 @@
+package datasetAPI
```
worth putting this straight into Carl's `publishing/datasetAPI/` dir, rather than here, since it's a new test?

gedge commented on datasetAPI/get_observation_test.go:
```diff
@@ -0,0 +1,306 @@
+				response.Value("dimensions").Object().Value("time").Object().Value("options").Array().Element(0).Object().Value("id").Equal("Aug-16")
+				response.Value("observation").Equal("117.9")
+				response.Value("links").Object().Value("dataset_metadata").Object().Value("href").String().Match("(.+)/datasets/" + datasetID + "/editions/" + edition + "/versions/1/metadata$")
+				response.Value("links").Object().Value("self").Object().Value("href").String().Match(".+/datasets/" + datasetID + "/editions/" + edition + "/versions/1/observations\\?aggregate=cpi1dim1G50100&geography=K02000001&time=Aug-16$") // ?aggregate=cpi1dim1S40403&geography=K02000001&time=Aug-16")
```
remove trailing comment

gedge commented on datasetAPI/get_observation_test.go:
```diff
@@ -0,0 +1,306 @@
+			Convey("When a request to get an observation for a version of a dataset", func() {
+				Convey("Then return status not found (404) with message `Edition not found`", func() {
+					datasetAPI.GET("/datasets/{id}/editions/{edition}/versions/2/observations", datasetID, edition).
+						WithQueryString("time=Aug-16&geography=K02000001&aggregate=cpi1dim1S40403").WithHeader(florenceTokenName, florenceToken).
```
I'd prefer these lines to be broken more, to make them more readable (I think github makes it even worse), so start new lines with
- `WithHeader`
- `Expect`
- `Status`
- `Body`

nshumoogum committed:
Check observation response contains unit of measure

gedge commented on datasetAPI/get_observation_test.go:
```diff
@@ -0,0 +1,306 @@
+					Convey("Then return status bad request (400) with a message", func() {
+						datasetAPI.GET("/datasets/{id}/editions/{edition}/versions/2/observations", datasetID, edition).WithHeader(florenceTokenName, florenceToken).
+							WithQueryString("geography=K02000001&aggregate=cpi1dim1S40403").
+							Expect().Status(http.StatusBadRequest).Body().Contains("Missing query parameters for the following dimensions: [time]")
```
would prefer to see >1 dimension missing, to test that the error shows multiple values

gedge commented on datasetAPI/get_version_test.go:
```diff
@@ -65,7 +65,10 @@ func TestSuccessfullyGetVersionOfADatasetEdition(t *testing.T) {
 
 			Convey("When an authenticated request including a valid download service token is made to get the published version", func() {
-				response := datasetAPI.GET("/datasets/{id}/editions/{edition}/versions/2", datasetID, edition).WithHeader(downloadServiceAuthToken, downloadServiceAuthTokenID).WithHeader(serviceAuthTokenName, serviceAuthToken).
+				headers := make(map[string]string)
```
perhaps put this as `authHeaders` shared at top of this file?

gedge commented on datasetAPI/get_versions_test.go:
```diff
@@ -303,6 +339,13 @@ func checkVersionResponse(datasetID, editionID, instanceID, edition string, resp
 	response.Value("version").Equal(1)
 }
 
+func checkPublicOrPrivateLinksDoNotExistInResponse(response *httpexpect.Object) {
```
ambiguous name: consider `checkNeitherPublicNorPrivateLinksExistInResponse`

nshumoogum committed:
Remove unnecessary capture group in string match

nshumoogum committed:
Use different bucket names in private/public href values

This means the tests will now fail if a bug causes private link to equal
public link and vice versa.

nshumoogum committed:
Check multiple dimensions missing when calling dataset observations
endpoint

Also break up lines to make them more readable and remove trailing
comments

nshumoogum committed:
Rename shared headers at top of test for reuse

Also change function name to be less ambiguous

nshumoogum committed:
Merge branch 'cmd-develop' into feature/dataset-observation-endpoint

nshumoogum committed:
Move dataset API observation endpoint tests to new public and web

Also update out of date README.md

gedge commented on README.md:
```diff
@@ -3,17 +3,37 @@ dp-api-tests
+If you're running applications using websysd ensure `InheritEnvironment`
+is set to true in the `websysd.json` file.
+
+The acceptance tests are split between running in publishing or web and result
```
pls to be marking-up:
```
The acceptance tests are split between running in *publishing* or *web* and result
```

gedge commented on README.md:
```diff
@@ -3,17 +3,37 @@ dp-api-tests
+
+To test services in publishing run the following command:
+
+`cd publishing; go test ./...`
```
```
`cd publishing && go test ./...`
```

gedge commented on README.md:
```diff
@@ -3,17 +3,37 @@ dp-api-tests
+in web and then run the following command:
 
-`go test ./...` - run all the tests
+`cd web; go test ./...`
```
```
`cd web && go test ./...`
```

gedge commented on publishing/datasetAPI/README.md:
````diff
@@ -0,0 +1,22 @@
+pass:
+
+```text
+mongo db
````
`mongodb` is one word

gedge commented on web/datasetAPI/README.md:
```diff
@@ -0,0 +1,21 @@
+
+### Getting started
+
+This package will test all endpoints that exist within the Dataset API
```
Perhaps `This package will test all endpoints that are exposed by the Dataset API when running in web/public subnet.` ?

nshumoogum committed:
Update observations endpoint tests

nshumoogum committed:
Merge branch 'feature/dataset-observations-endpoint' into feature/dataset-observation-endpoint

nshumoogum committed:
Update root, web and publishing README.md's

gedge commented on publishing/datasetAPI/get_observation_test.go:
```diff
@@ -89,19 +90,73 @@ func TestSuccessfullyGetObservationForVersion(t *testing.T) {
+				response.Value("observations").Array().Length().Equal(137)
+
+				// check two observations in observations array
+				for i := 0; i < len(response.Value("observations").Array().Iter()); i++ {
```
this `for` loop is a tad broken...
1. use `for _, observation := range response.Value("observations").Array().Iter() {`
2. put `count := 0` _outside_ the loop
3. test that `count` is 2 *after* the loop is done

nshumoogum replied:
I will use the range for loop and remove counter as it appears to be eating up time then gaining time.

gedge commented on publishing/datasetAPI/get_observation_test.go:
```diff
@@ -89,19 +90,73 @@ func TestSuccessfullyGetObservationForVersion(t *testing.T) {
+						count++
+					}
+
+					if count == 2 {
```
can get rid of this

gedge commented on web/datasetAPI/get_observation_test.go:
```diff
@@ -44,26 +44,78 @@ func TestSuccessfullyGetObservationForVersion(t *testing.T) {
+				response.Value("observations").Array().Length().Equal(137)
+
+				// check two observations in observations array
+				for i := 0; i < len(response.Value("observations").Array().Iter()); i++ {
```
same problem as above

nshumoogum committed:
Remove counter to assist breaking out of for loop

It is actually more time consuming having the counter than looping
through all observations (137) in test.

nshumoogum committed:
Check two observations were seen when testing response body from
observations endpoint

If the test data changes and these two observations are removed test
will still pass even if the data structure of the observations object
changes, by putting a check on how many observations are seen we prevent
this being the case.

nshumoogum committed:
Check the number of unique observations (based on code) matches
expectation

nshumoogum committed:
Merge branch 'cmd-develop' into feature/dataset-observation-endpoint
